src/Cake.Core/IO/NuGet/AssemblyFilePathFrameworkNameParser.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Linq;
+using System.Runtime.Versioning;
+
+namespace Cake.Core.IO.NuGet
+{
+    /// <summary>
+    /// Represents an object that parses the segments of a filepath for a .Net framework name
+    /// </summary>
+    public class AssemblyFilePathFrameworkNameParser : IAssemblyFilePathFrameworkNameParser
+    {
+        private static readonly FrameworkName _unsupportedFrameworkName = new FrameworkName("Unsupported", new Version());
+        private readonly IFrameworkNameParser _frameworkNameParser;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AssemblyFilePathFrameworkNameParser"/> class.
+        /// </summary>
+        /// <param name="frameworkNameParser">The framework name parser.</param>
+        public AssemblyFilePathFrameworkNameParser(IFrameworkNameParser frameworkNameParser)
+        {
+            _frameworkNameParser = frameworkNameParser;
+        }
+        
+        /// <summary>
+        /// Parses the framework name from assembly file path.
+        /// </summary>
+        /// <param name="path">The assembly file path.</param>
+        /// <returns>the parsed framework name -- null when path contains no folders</returns>
+        public FrameworkName Parse(FilePath path)
+        {
+            if (path == null)
+            {
+                throw new ArgumentNullException("path");
+            }
+
+            // The path for a reference might look like this for assembly foo.dll:            
+            // foo.dll
+            // sub\foo.dll
+            // {FrameworkName}{Version}\foo.dll
+            // {FrameworkName}{Version}\sub1\foo.dll
+            // {FrameworkName}{Version}\sub1\sub2\foo.dll
+            return ParseFromDirectoryPath(path.GetDirectory());
+        }
+
+        private FrameworkName ParseFromDirectoryPath(DirectoryPath path)
```

Comment:
We've been trying to minimizing recursion in the script base as much as we can, so it's more a preference than a rule. 

Not sure I understand the code completely, but I was thinking of something like this.

```csharp
private FrameworkName ParseFromDirectoryPath(DirectoryPath path)
{
  var queue = new Queue<string>(path.Segments);
  while(queue.Count > 0)
  {
    var current = queue.Dequeue();
    var parsedFxName = _frameworkNameParser.ParseFrameworkName(targetFrameworkString);
    if (parsedFxName != _unsupportedFrameworkName)
    {
      return parsedFxName;
    }
  }
  return null;
}
```